proposals/0000-support-pun-free-code.md

```diff
@@ -0,0 +1,510 @@
+(to keep compatibility with existing code) and introduces `type` syntactic marker.
+
+However, thanks to *Syntactic Unification Principle* (adhered by #281), if
+the user chooses to not use punning, there is no need to use this syntactic
+marker, resulting in less context-dependent and syntax cluttered code.
+
+Unfortunately, GHC has no good support for such style of programming:
+
+* There is no way for the user to know if some code uses punning without
+  manually reviewing it, in other words, usage of punning is silent.
+
+* There is no way to avoid punning if the user imports the code that makes the
+  use of it.
+  Consider this import statement: `import Data.Proxy`. It is impossible to
+  distinguish between `Proxy` the type and `Proxy` the data constructor.
+
+#### `'` syntax clash in `-XDataKinds` and `-XTemplateHaskell`
+
+`'` prefix syntax is used both by `-XDataKinds`, to promote a constructor,
+and in `-XTemplateHaskell`, to refer to a term-level identifier. For
+type-level identifiers `-XTemplateHaskell` uses `''`.
+
+This not only results in ambiguity when both of the extensions are enabled, it violates
+the *Syntactic Unification Principle* as users have to use `'` in
+`-XTemplateHaskell` for terms and `''` for types whether they use punning or
+not.
+
+#### Beginner confusion
+
+Even without Dependent Haskell an argument can be made for pun-free code:
+punning can be a source of confusion for beginners. The difference between
+the terms namespace and the types namespace can be hard to understand at first,
+especially when things like `()` or `[a]` are used ([as seen in this StackOverflow question](https://stackoverflow.com/questions/16892570/what-is-in-haskell-exactly)).
+
+### Solution Overview
+
+#### Deprecating `'` in `-XDataKinds` and `''` in `-XTemplateHaskell`
+
+In order to solve the violation of the *Syntactic Unification Principle*
+introduced by `-XTemplateHaskell` we propose to deprecate `'` syntax in
+`-XDataKinds` and to deprecate `''` syntax in `-XTemplateHaskell`. Instead, `'`
+is only used in `-XTemplateHaskell` to refer to both term-level and type-level
+identifiers.
+
+#### Namespace qualified imports
+
+To help programmers deal with the imported code that uses punning we propose
+to introduce namespace qualified import syntax, guarded behind
+`-XExplicitNamespaces` extension. The syntax introduces two markers for
+qualified imports, `data` and `type`:
+
+```haskell
+import Data.Proxy type qualified as T   -- import only the type namespace
+import Data.Proxy data qualified as D   -- import only the data namespace
+```
+
+#### Module aliasing
+
+Deprecating `'` from `-XDataKinds` renders it impossible to refer to punned data
+constructors in type context defined within the same module and makes `'`
+in `-XTemplateHaskell` ambiguous. To mitigate this we propose a new extension,
+`-XModuleAliases`, that lets users write aliases to the current module and each
+of its namespaces in the module header:
```

Comment:
I am not convinced that this is hammer is right-sized for this problem, but I may be missing something, right?

The use case here is code that wants to use punning, even though we disabled the use of `''`, right? So I wonder: 
* Do we have to remove support for `'`? Can't we keep it as long as Haskell has two namespaces (possibly forever)?
* If yes, is disambiguation with `(type …)` and `(data …)` not sufficient? (And more readable and easier to learn).

I see why for importing punning modules you want something more concise. But as a clutch for code that has to provide punning (and use TH etc.), it seems the existing solution of name space qualifiers should be sufficient?

(Or do we only have `(type …)` but not `(data …)`? Then we should probably fix this inconsistency, should we?)

Reply:
I've been thinking about your comment and I think you're right.

It seems to me that the best way to move forward is to not touch DataKinds and change TH to only use `''` which will default to type namespace (but will fallback to data namespace if the name wasn't found in the type namespace). And then for punning add something like `''(data T)` and `''(type T)` (for consistency).

This way `''` keeps compatibility with the way `''` works now, which is good, especially given that perhaps the most common use case for TH is to generate lens/json/etc instances, and we also don't have to touch DataKinds.

Reply:
Why again even touch `'` and `''` in the first place? In a pun-free world, theyd be equivalent. In other code, they start with a different default, but you can use the namespace overrides of course.

Maybe in a decade or sk, when and if pun-free code becomes common, we can think about removing this small syntactic redundancy. Until then, why touch this?

Reply:
I agree with @nomeata here. I'm not convinced we need to touch Template Haskell here -- except perhaps to make it so that `'` and `''` look up in the alternative namespace when the primary lookup fails -- and to support some means of disambiguation. Both of these would be fully backward compatible.

We can also use the choice between term-syntax and type-syntax to interpret a `'`: in term-syntax, it means a TH name quote. In type-syntax, it means a (deprecated) namespace selection.

Reply:
After the discussion yesterday I think I'll leave `'` and `''` stuff to a separate proposal. This is the most unclear and controversial change of this proposal and I don't want to block `-Wpuns`/`-Wpun-bindings` and namespace qualifiers with it.

Reply:
After the discussion yesterday I think I'll leave `'` and `''` stuff to a separate proposal. This is the most unclear and controversial change of this proposal and I don't want to block `-Wpuns`/`-Wpun-bindings` and namespace qualifiers with it.